File: backend/sdk/python/waas_sdk/verify.py
"""
WaaS Webhook Signature Verification - Standalone, zero-dependency module.

Usage:
    from waas_sdk.verify import WebhookVerifier

    verifier = WebhookVerifier("whsec_your_secret")
    verifier.verify(payload_bytes, signature_header, timestamp_header)

Framework middleware:
    # Django
    from waas_sdk.verify import django_middleware
    MIDDLEWARE = ['waas_sdk.verify.django_middleware']

    # Flask/Express-like
    from waas_sdk.verify import verify_request
"""

import hashlib
import hmac
import math
import time as _time


class VerificationError(Exception):
    """Base class for signature verification errors."""
    pass


class MissingSignatureError(VerificationError):
    """Raised when the signature header is missing."""
    pass


class InvalidSignatureError(VerificationError):
    """Raised when the signature does not match."""
    pass


class TimestampExpiredError(VerificationError):
    """Raised when the timestamp is outside the tolerance window."""
    pass


class WebhookVerifier:
    """Verifies WaaS webhook signatures using HMAC-SHA256.

    Args:
        secret: Primary signing secret.
        secrets: Additional secrets for key rotation.
        timestamp_tolerance: Maximum allowed age in seconds (default: 300).
    """

    SIGNATURE_HEADER = "X-WaaS-Signature"
    TIMESTAMP_HEADER = "X-WaaS-Timestamp"

    def __init__(self, secret: str, secrets: list = None, timestamp_tolerance: int = 300):
        self._secrets = [secret] + (secrets or [])
        self._timestamp_tolerance = timestamp_tolerance

    def verify(self, payload: bytes, signature: str, timestamp: str = "") -> bool:
        """Verify a webhook signature.

        Args:
            payload: Raw request body bytes.
            signature: Value of X-WaaS-Signature header (format: v1=<hex>).
            timestamp: Value of X-WaaS-Timestamp header (Unix seconds string).

        Returns:
            True if signature is valid.

        Raises:
            MissingSignatureError: If signature is empty.
            InvalidSignatureError: If signature doesn't match any secret.
            TimestampExpiredError: If timestamp is outside tolerance.
        """
        if not signature:
            raise MissingSignatureError("Missing signature header")

        self._verify_timestamp(timestamp)

        sig_bytes = self._parse_signature(signature)
        signed_payload = f"{timestamp}.{payload.decode('utf-8') if isinstance(payload, bytes) else payload}"

        for secret in self._secrets:
            expected = hmac.new(
                secret.encode("utf-8"),
                signed_payload.encode("utf-8"),
                hashlib.sha256,
            ).digest()
            if hmac.compare_digest(sig_bytes, expected):
                return True

        raise InvalidSignatureError("Signature does not match")

    def sign(self, payload: bytes, timestamp: str) -> str:
        """Generate a signature for testing purposes."""
        signed_payload = f"{timestamp}.{payload.decode('utf-8') if isinstance(payload, bytes) else payload}"
        mac = hmac.new(
            self._secrets[0].encode("utf-8"),
            signed_payload.encode("utf-8"),
            hashlib.sha256,
        ).hexdigest()
        return f"v1={mac}"

    def _verify_timestamp(self, timestamp: str):
        if not timestamp or self._timestamp_tolerance <= 0:
            return
        try:
            ts = int(timestamp)
        except ValueError as e:
            raise VerificationError(f"Malformed timestamp: {e}")

        diff = abs(_time.time() - ts)
        if diff > self._timestamp_tolerance:
            raise TimestampExpiredError(
                f"Timestamp expired: {diff:.0f}s > {self._timestamp_tolerance}s tolerance"
            )

    @staticmethod
    def _parse_signature(header: str) -> bytes:
        parts = header.split("=", 1)
        if len(parts) != 2:
            raise VerificationError("Malformed signature header")
        return bytes.fromhex(parts[1])


def verify_request(secret: str, payload: bytes, headers: dict) -> bool:
    """Convenience function to verify a webhook request.

    Args:
        secret: Signing secret.
        payload: Raw request body.
        headers: Request headers dict.
    """
    v = WebhookVerifier(secret)
    sig = headers.get(WebhookVerifier.SIGNATURE_HEADER, "")
    ts = headers.get(WebhookVerifier.TIMESTAMP_HEADER, "")
    return v.verify(payload, sig, ts)


def django_middleware(get_response):
    """Django middleware for WaaS webhook signature verification.

    Configure WAAS_WEBHOOK_SECRET in Django settings.

    Usage in settings.py:
        MIDDLEWARE = [..., 'waas_sdk.verify.django_middleware']
        WAAS_WEBHOOK_SECRET = 'whsec_...'
    """
    from django.conf import settings
    from django.http import HttpResponseForbidden

    secret = getattr(settings, "WAAS_WEBHOOK_SECRET", "")

    def middleware(request):
        if request.method == "POST" and request.content_type == "application/json":
            sig = request.META.get("HTTP_X_WAAS_SIGNATURE", "")
            ts = request.META.get("HTTP_X_WAAS_TIMESTAMP", "")
            if sig:
                try:
                    v = WebhookVerifier(secret)
                    v.verify(request.body, sig, ts)
                except VerificationError as e:
                    return HttpResponseForbidden(f"Webhook verification failed: {e}")
        return get_response(request)

    return middleware

File: backend/sdk/ruby/lib/waas/verify.rb
# frozen_string_literal: true

require "openssl"

module Waas
  # Standalone, zero-dependency WaaS webhook signature verification.
  #
  # Usage:
  #   verifier = Waas::WebhookVerifier.new("whsec_your_secret")
  #   verifier.verify!(payload, signature_header, timestamp_header)
  #
  # Rails middleware:
  #   config.middleware.use Waas::WebhookVerifier::RackMiddleware, "whsec_..."
  #
  class WebhookVerifier
    SIGNATURE_HEADER = "X-WaaS-Signature"
    TIMESTAMP_HEADER = "X-WaaS-Timestamp"
    DEFAULT_TOLERANCE = 300

    class VerificationError < StandardError; end
    class MissingSignatureError < VerificationError; end
    class InvalidSignatureError < VerificationError; end
    class TimestampExpiredError < VerificationError; end

    # @param secret [String] Primary signing secret
    # @param secrets [Array<String>] Additional secrets for key rotation
    # @param timestamp_tolerance [Integer] Max allowed age in seconds
    def initialize(secret, secrets: [], timestamp_tolerance: DEFAULT_TOLERANCE)
      @secrets = [secret] + secrets
      @timestamp_tolerance = timestamp_tolerance
    end

    # Verify a webhook signature.
    #
    # @param payload [String] Raw request body
    # @param signature [String] Value of X-WaaS-Signature header
    # @param timestamp [String] Value of X-WaaS-Timestamp header
    # @return [true] if valid
    # @raise [VerificationError] if invalid
    def verify!(payload, signature, timestamp = "")
      raise MissingSignatureError, "Missing signature header" if signature.nil? || signature.empty?

      verify_timestamp!(timestamp)

      sig_bytes = parse_signature(signature)
      signed_payload = "#{timestamp}.#{payload}"

      @secrets.each do |secret|
        expected = OpenSSL::HMAC.digest("SHA256", secret, signed_payload)
        return true if secure_compare(sig_bytes, expected)
      end

      raise InvalidSignatureError, "Signature does not match"
    end

    # Generate a signature for testing.
    def sign(payload, timestamp)
      signed_payload = "#{timestamp}.#{payload}"
      mac = OpenSSL::HMAC.hexdigest("SHA256", @secrets.first, signed_payload)
      "v1=#{mac}"
    end

    private

    def verify_timestamp!(timestamp)
      return if timestamp.nil? || timestamp.empty? || @timestamp_tolerance <= 0

      ts = Integer(timestamp)
      diff = (Time.now.to_i - ts).abs
      if diff > @timestamp_tolerance
        raise TimestampExpiredError, "Timestamp expired: #{diff}s > #{@timestamp_tolerance}s tolerance"
      end
    rescue ArgumentError => e
      raise VerificationError, "Malformed timestamp: #{e.message}"
    end

    def parse_signature(header)
      parts = header.split("=", 2)
      raise VerificationError, "Malformed signature header" if parts.length != 2

      [parts[1]].pack("H*")
    end

    def secure_compare(a, b)
      return false unless a.bytesize == b.bytesize

      OpenSSL.fixed_length_secure_compare(a, b)
    rescue StandardError
      # Fallback for older Ruby versions
      a == b
    end

    # Rack/Rails middleware for webhook signature verification.
    #
    # Usage in Rails:
    #   config.middleware.use Waas::WebhookVerifier::RackMiddleware, "whsec_..."
    class RackMiddleware
      def initialize(app, secret, **options)
        @app = app
        @verifier = WebhookVerifier.new(secret, **options)
      end

      def call(env)
        return @app.call(env) unless env["REQUEST_METHOD"] == "POST"

        signature = env["HTTP_X_WAAS_SIGNATURE"]
        return @app.call(env) if signature.nil? || signature.empty?

        timestamp = env["HTTP_X_WAAS_TIMESTAMP"] || ""
        body = env["rack.input"].read
        env["rack.input"].rewind

        begin
          @verifier.verify!(body, signature, timestamp)
          @app.call(env)
        rescue VerificationError => e
          [401, { "Content-Type" => "application/json" }, [%({"error":"#{e.message}"})]]
        end
      end
    end
  end
end
